src/org/exist/xquery/functions/transform/EXistURIResolver.java

```diff
@@ -0,0 +1,176 @@
+
+        if (numSimplifiedComponents == 0) {
+            return "/";
+        }
+
+        final StringBuilder b = new StringBuilder(path.length());
+        for(int x = 0; x < numSimplifiedComponents; x++) {
+            b.append("/").append(simplifiedComponents[x]);
+        }
+
+        if (path.endsWith("/")) {
+            b.append("/");
+        }
+
+        return b.toString();
+    }
+
+    @Override
+    public Source resolve(final String href, String base) throws TransformerException {
+        String path;
+
+        if (href.isEmpty()) {
+            path = base;
+        } else {
+            URI hrefURI = null;
+            try {
+                hrefURI = new URI(href);
+            } catch (final URISyntaxException e) {
+            }
+            if (hrefURI != null && hrefURI.isAbsolute()) {
+                path = href;
+            } else {
+                if (href.startsWith("/")) {
+                    path = href;
+                } else if (href.startsWith(XmldbURI.EMBEDDED_SERVER_URI_PREFIX)) {
+                    path = href.substring(XmldbURI.EMBEDDED_SERVER_URI_PREFIX.length());
+                } else if (base == null || base.length() == 0) {
+                    path = basePath + "/" + href;
+                } else {
+                    // Maybe base never contains this prefix?  Check to be sure.
+                    if (base.startsWith(XmldbURI.EMBEDDED_SERVER_URI_PREFIX)) {
+                        base = base.substring(XmldbURI.EMBEDDED_SERVER_URI_PREFIX.length());
+                    }
+                    path = base.substring(0, base.lastIndexOf("/") + 1) + href;
+                }
+            }
+        }
+        LOG.debug("Resolving path " + href + " with base " + base + " to " + path);// + " (URI = " + uri.toASCIIString() + ")");
+
+        if (path.startsWith("/")) {
+            path = normalizePath(path);
+            return databaseSource(path);
+        } else {
+            return urlSource(path);
+        }
+    }
+
+    private Source urlSource(final String path) throws TransformerException {
+        try {
+            final URL url = new URL(path);
+            return new StreamSource(url.openStream());
+        } catch (final IOException e) {
+            throw new TransformerException(e.getMessage(), e);
+        }
+    }
+
+    private Source databaseSource(final String path) throws TransformerException {
+        final XmldbURI uri = XmldbURI.create(path);
+
+        final DocumentImpl doc;
+        try {
+            doc = broker.getResource(uri, Permission.READ);
+            if (doc == null) {
+                LOG.error("Document " + path + " not found");
+                throw new TransformerException("Resource " + path + " not found in database.");
+            }
+
+            final Source source;
+            if (doc instanceof BinaryDocument) {
+                final Path p = broker.getBinaryFile((BinaryDocument) doc);
+                source = new StreamSource(p.toFile());
```

Comment:
Yes we have direct access to binary files in WEB-INF/data/fs for the purpose of reading